File: app/src/main/java/ru/nikshlykov/englishwordsapp/db/subgroup/SubgroupDao.java
package ru.nikshlykov.englishwordsapp.db.subgroup;

import android.database.Cursor;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Update;

import java.util.List;

@Dao
public interface SubgroupDao {

    long GROUP_FOR_NEW_SUBGROUPS_ID = -1L;

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    long insert(Subgroup subgroup);

    @Update
    int update(Subgroup subgroup);

    @Delete
    int delete(Subgroup subgroup);

    @Query("SELECT * FROM Subgroups WHERE _id = :id")
    Subgroup getSubgroupById(long id);

    @Query("SELECT * FROM Subgroups WHERE _id = :id")
    LiveData<Subgroup> getLiveDataSubgroupById(long id);
/*
    @Query("SELECT * FROM Subgroups")
    LiveData<List<Subgroup>> getAllSubgroups();*/

    @Query("SELECT * FROM Subgroups WHERE groupId = :groupId")
    List<Subgroup> getSubgroupsFromGroup(long groupId);

    /*@Query("SELECT * FROM Subgroups WHERE IsStudied = 1")
    Subgroup[] getStudiedSubgroups();*/

    @Query("SELECT * FROM Subgroups ORDER BY _id LIMIT 1")
    Subgroup getSubgroupWithMinId();

    @Query("SELECT * FROM Subgroups WHERE groupId = " + GROUP_FOR_NEW_SUBGROUPS_ID)
    Subgroup[] getCreatedByUserSubgroups();
}

File: app/src/main/java/ru/nikshlykov/englishwordsapp/db/subgroup/SubgroupDao.kt
package ru.nikshlykov.englishwordsapp.db.subgroup

import androidx.lifecycle.LiveData
import androidx.room.*
import ru.nikshlykov.englishwordsapp.db.group.Group

@Dao
interface SubgroupDao {

  @Insert(onConflict = OnConflictStrategy.IGNORE)
  suspend fun insert(subgroup: Subgroup): Long

  @Update
  suspend fun update(subgroup: Subgroup): Int

  @Delete
  suspend fun delete(subgroup: Subgroup): Int

  @Query("SELECT * FROM Subgroups WHERE _id = :id")
  fun getSubgroupById(id: Long): Subgroup

  @Query("SELECT * FROM Subgroups WHERE _id = :id")
  suspend fun getSubgroupByIdSuspend(id: Long): Subgroup

  @Query("SELECT * FROM Subgroups WHERE groupId = :groupId")
  suspend fun getSubgroupsFromGroup(groupId: Long): List<Subgroup>

  @Query("SELECT * FROM Subgroups ORDER BY _id LIMIT 1")
  suspend fun subgroupWithMinId(): Subgroup

  @Query("SELECT * FROM Subgroups WHERE groupId = ${Group.GROUP_FOR_NEW_SUBGROUPS_ID}")
  fun createdByUserSubgroups(): Array<Subgroup>
}